File: python/src/scientific_workflow/project.py
"""Focused accessors for the REQUIRED standard study layout and launch environment.

These helpers do not discover relocated files, activate environments, change
working directories, or create output. Program snapshots contain resolved values.
"""
import json
import os
from pathlib import Path


class ProjectLayoutError(ValueError):
    """A required standard path or snapshot is absent or malformed."""


def _environment_path(variable: str, *, directory: bool) -> Path:
    value = os.environ.get(variable)
    if not value:
        raise ProjectLayoutError(f"missing {variable}; run through Workflow's REQUIRED standard study layout")
    path = Path(value)
    if not path.is_absolute() or not (path.is_dir() if directory else path.is_file()):
        raise ProjectLayoutError(f"expected {'directory' if directory else 'file'} at {path} from {variable}; preserve Workflow's REQUIRED standard layout")
    return path


def project_root() -> Path:
    """Return the launched program's WORKFLOW_PROJECT_ROOT."""
    return _environment_path("WORKFLOW_PROJECT_ROOT", directory=True)


def output_directory() -> Path:
    """Return the current program's writable standard artifacts directory."""
    return _environment_path("WORKFLOW_TASK_OUTPUT", directory=True)


def study_path(root: str | Path) -> Path:
    """Require <root>/wf_configs/study.json, without parsing source configuration."""
    path = Path(root) / "wf_configs" / "study.json"
    if not path.is_file():
        raise ProjectLayoutError(f"expected {path}; REQUIRED layout: <study>/wf_configs/study.json")
    return path


def parameters(section: str | None = None, *, snapshot: str | Path | None = None) -> object:
    """Read resolved parameters, optionally selecting one exact top-level key.

    Omit snapshot only inside a Workflow-launched program. This does not read
    unresolved wf_configs/parameters.json or implement a second resolution graph.
    """
    path = Path(snapshot) if snapshot is not None else _environment_path("WORKFLOW_CONFIG_PATH", directory=False)
    try:
        value = json.loads(path.read_text(encoding="utf-8"))["config"]["parameters.json"]
        if not isinstance(value, dict):
            raise ValueError("parameters.json must contain an object")
        return value if section is None else value[section]
    except (OSError, ValueError, KeyError, TypeError) as error:
        raise ProjectLayoutError(f"cannot read resolved parameters {section!r} at {path}; expected Workflow's standard workflow-config.json snapshot: {error}") from error

File: rust/src/task/project.rs
//! Accessors for Workflow's required project layout and resolved program snapshots.
use serde::de::DeserializeOwned;
use std::path::{Path, PathBuf};
use thiserror::Error;

/// A missing standard path, environment variable, or invalid resolved snapshot.
#[derive(Debug, Error)]
#[error("{message}; preserve Workflow's required study layout")]
pub struct ProjectLayoutError {
    message: String,
    #[source]
    source: Option<Box<dyn std::error::Error + Send + Sync>>,
}
fn failure(message: String) -> ProjectLayoutError {
    ProjectLayoutError {
        message,
        source: None,
    }
}
fn environment_path(variable: &str, directory: bool) -> Result<PathBuf, ProjectLayoutError> {
    let value = std::env::var_os(variable).ok_or_else(|| failure(format!("missing {variable}")))?;
    let path = PathBuf::from(value);
    if !path.is_absolute()
        || !(if directory {
            path.is_dir()
        } else {
            path.is_file()
        })
    {
        return Err(failure(format!(
            "expected {} at {} from {variable}",
            if directory { "directory" } else { "file" },
            path.display()
        )));
    }
    Ok(path)
}
/// Returns `WORKFLOW_PROJECT_ROOT` for a launched program; does not change cwd.
pub fn project_root() -> Result<PathBuf, ProjectLayoutError> {
    environment_path("WORKFLOW_PROJECT_ROOT", true)
}
/// Returns `WORKFLOW_TASK_OUTPUT`, the existing writable artifacts directory.
pub fn output_directory() -> Result<PathBuf, ProjectLayoutError> {
    environment_path("WORKFLOW_TASK_OUTPUT", true)
}
/// Requires `<root>/wf_configs/study.json` and returns its path without parsing it.
pub fn study_path(root: &Path) -> Result<PathBuf, ProjectLayoutError> {
    let path = root.join("wf_configs/study.json");
    if !path.is_file() {
        return Err(failure(format!("expected {}", path.display())));
    }
    Ok(path)
}
/// Loads resolved parameters from `WORKFLOW_CONFIG_PATH` and deserializes `T`.
///
/// `section` selects one exact top-level key; `None` selects all parameters.
/// This does not read unresolved project declarations or repeat Config resolution.
pub fn parameters<T: DeserializeOwned>(section: Option<&str>) -> Result<T, ProjectLayoutError> {
    parameters_from_snapshot(&environment_path("WORKFLOW_CONFIG_PATH", false)?, section)
}
/// Loads resolved parameters from an explicit standard `workflow-config.json` file.
///
/// The file must contain `config["parameters.json"]`. Filesystem/JSON/typed
/// deserialization errors retain their cause. A failed read returns no value.
pub fn parameters_from_snapshot<T: DeserializeOwned>(
    path: &Path,
    section: Option<&str>,
) -> Result<T, ProjectLayoutError> {
    let read = || -> Result<T, Box<dyn std::error::Error + Send + Sync>> {
        let bytes = std::fs::read(path)?;
        let snapshot: serde_json::Value = serde_json::from_slice(&bytes)?;
        let parameters = snapshot
            .get("config")
            .and_then(|v| v.get("parameters.json"))
            .filter(|v| v.is_object())
            .ok_or("expected config[parameters.json] object")?;
        let selected = match section {
            Some(key) => parameters
                .get(key)
                .ok_or_else(|| format!("missing parameter section {key:?}"))?,
            None => parameters,
        };
        Ok(serde_json::from_value(selected.clone())?)
    };
    read().map_err(|source| ProjectLayoutError {
        message: format!(
            "cannot read resolved parameters at {}: {source}",
            path.display()
        ),
        source: Some(source),
    })
}
